Add script to generate and upload schemas

fbertsch opened the pull request.

fbertsch commented:
There's still a few TODOs:

- Will this script create a tarball and push that to GCS, or will jenkins see the new push to MPS and create its own?
- We need to get the proper credentials and bucket for GCS public repos
- We need to setup credentials to let this script push to MPS

fbertsch commented:
@whd, you seem most likely to have opinions on those todos listed above.

fbertsch commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,75 @@
+
+# 0. Install dependencies
+
+cargo install --git https://github.com/acmiyaguchi/jsonschema-transpiler --branch 0.2 --force
```
@acmiyaguchi is this still how the transpiler should be retrieved?

acmiyaguchi replied:
Yes, though this will change very soon. I'm moving the repository to the mozilla org either tomorrow or next week, and making master the default branch. I'm also trying to generate binaries for cross platform linux use, but the infrastructure for that isn't set up (and installing cargo isn't too bad anyways).

fbertsch replied:
Great, keep us updated. The cross-platform binaries could be useful for MPS testing, but aren't required here.

acmiyaguchi replied:
I moved the repo to the mozilla org and [published a package](https://crates.io/crates/jsonschema-transpiler). This can now be `cargo install jsonschema-transpiler`.

fbertsch replied:
Great news! Updating now.

fbertsch commented on bin/metadata_merge:
```diff
@@ -0,0 +1,57 @@
+    ),
+    required=True
+)
+def main(metadata, schema):
```
@jklukas is this how you envisioned metadata being merged?

jklukas replied:
Yes, this looks entirely reasonable.

fbertsch commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,75 @@
+
+git clone https://www.github.com/mozilla-services/mozilla-pipeline-schemas
+cd mozilla-pipeline-schemas/$SCHEMAS_DIR
+git checkout ping-metadata
```
When we merge the GCS metadata, we'll checkout `dev` instead.

whd commented:
>     * Will this script create a tarball and push that to GCS, or will jenkins see the new push to MPS and create its own?

The latter.

>     * We need to get the proper credentials and bucket for GCS public repos

Absent a compelling reason, this should also happen on jenkins (as triggered by the MPS push).

>     * We need to setup credentials to let this script push to MPS

I will work on setting this up for airflow.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,75 @@
+cd ../
+
+rm -rf templates tests validation
+find . -not -name "*.schema.json" -type f | xargs git add
```
Does this flow correctly do nothing when no changes have been made to the schemas?

fbertsch replied:
I've confirmed that there is no commit pushed when there are no changes.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,75 @@
+
+# 2. Remove all non-json schemas (e.g. parquet)
+
+find . -not -name "*.schema.json" -type f | xargs rm
```
Is this necessary? I can image a future in which we have a unified branch (master) that is updated by both humans and machines, where parquet schemas are simply ignored in the GCP pipeline.

fbertsch replied:
Not really necessary, but I would prefer to only have the relevant schemas available. Why don't we add the parquet schemas back in later if they become useful?

fbertsch replied:
Basically my goal is to have an easy way of saying "what is running on prod"? And the parquet schemas are not included in that (for GCP).

whd replied:
This is fine for now, since the generated branch will be gcp-only. I'll want to revisit this in the hypothetical future where we have a single branch. Since the responsibility of this script will not include uploading to GCS the deletion of parquetmr files could still happen farther down the pipeline.

whd commented on Dockerfile:
```diff
@@ -19,18 +19,31 @@ RUN apt-get update && \
 USER ${USER_ID}
+
+RUN git config --global user.name "Generated Schema Creator"
+RUN git config --global user.email "telemetry-alerts@mozilla.com"
```
Now that we've generated creds for this user, this should be `dataops+pipeline-schemas@mozilla.com`. Still TBD how we're going to pass in the credentials.

haroldwoo replied:
ref https://bugzilla.mozilla.org/show_bug.cgi?id=1547333

fbertsch replied:
Updated.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,75 @@
+git checkout $MPS_BRANCH || git checkout -b $MPS_BRANCH
+git commit -a -m "Auto-push from schema generation"
+
+git push --repo https://name:password@bitbucket.org/name/repo.git
```
Per https://bugzilla.mozilla.org/show_bug.cgi?id=1547333#c2 I think the easiest thing to do here is to set the ssh key via an env var and have this script write the contents of that var to `~/.ssh/id_rsa`.

fbertsch replied:
I've updated this, but testing locally I still needed to input a password. The only way I could force push to github was using a personal access token.

haroldwoo replied:
im not sure about bitbucket, but github requires pushing to the ssh url rather than https url for password less

fbertsch replied:
I did make that change, see the updated commit.

fbertsch commented:
I've added a whitelist, and fixed the cherry-picking. Testing led to this branch:

https://github.com/mozilla-services/mozilla-pipeline-schemas/tree/generated-schemas

However, I've yet to get it to work pushing using the ssh auth.

fbertsch commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,105 @@
+
+cd ../
+
+rm -rf templates tests validation
```
We could consider removing the remaining files in the git repository, e.g. `CMakeLists.txt`, `Dockerfile`, etc. They aren't necessarily useful in the read-only git branch.

whd replied:
I couldn't find an emoji for "this seems reasonable for a generated read-only git branch".

whd commented:
I've tested the generated-schemas branch on the jenkins infra and it mostly appears to work. The following issues still arise:

The `.bq` schemas aren't valid. This is a combination of https://github.com/mozilla/jsonschema-transpiler/issues/51 and (I think) the top-level `fields` being unnamed. @acmiyaguchi told me he's fixing at least the former of these.

The whitelist appears to work, but as a result a very limited subset of bq schemas are generated (e.g. no activity stream, webpagetest etc). Was this for testing purposes, or is bq table generation for less fancy schemas supposed to be incorporated into dev directly? Perhaps a blacklist would be better.

fbertsch commented:
@whd I added the whitelist so we could do a slow roll-out of schemas, while keeping the branch to contain only the ones that are deployed. All BQ table generation should be done during this build step.

whd commented:
@fbertsch ok, can you whitelist activity-stream and eng-workflow (bmobugs only)? These (and core v9/10) are the current tables we're generating in GCP so having them in the first round of automated schemas would be ideal.

jklukas commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,102 @@
+mozilla-schema-generator generate-glean-ping --out-dir . --pretty
+mozilla-schema-generator generate-main-ping --out-dir ./telemetry --pretty --split
+
+# 3a. Keep only whitelisted schemas
```
"blacklist" and "whitelist" are going the way of "slave"/"master" terminology. If we can avoid them, I think we should. I've found "allowlist" and "blocklist" to be suitable alternatives.

fbertsch replied:
Good catch - I had never realized this. Updated!

jklukas commented:
> can you whitelist activity-stream and eng-workflow

I want to be careful here, especially with activity-stream, that we don't accidentally recreate tables and drop existing data. As long as the logic here doesn't drop tables, this should be okay.

I've already updated the activity-stream, bmobugs, and core tables "by hand" to include the new metadata fields (and to relax the previously required metadata.document_id to nullable, since it is no longer being sent). These may throw errors when we try to update schema since they contain old metadata fields not in the json schemas.

After I validate that the new metadata fields are flowing properly, I plan to update these tables to drop the old fields (tomorrow (Wednesday), if all goes well).

whd commented:
> > can you whitelist activity-stream and eng-workflow
> 
> I want to be careful here, especially with activity-stream, that we don't accidentally recreate tables and drop existing data. As long as the logic here doesn't drop tables, this should be okay.

The hand-crafted tables have a `_dev` suffix on them, whereas the managed ones do not, no tables are dropped.  A separate config change needs to be applied at https://github.com/mozilla-services/cloudops-infra/blob/ingestion/projects/beam/tf/modules/resources/dataflow_jobs/bigquery.tf#L30 to point the dataflow jobs at the new managed tables.

jklukas commented:
Good point that we're transitioning to the non-dev namespaces now. I think this sounds good. Once we change the job to point at the new namespaces, I can work on backfilling the `dev` activity-stream tables to the new ones.

fbertsch commented:
@whd I've updated the whitelist and made the BQ output conform to expectations. You can check the current output at https://github.com/mozilla-services/mozilla-pipeline-schemas/tree/generated-schemas

whd commented:
Ok, I'm now seeing:

    * google_bigquery_table.telemetry_core_v5: googleapi: Error 400: The field specified for time partitioning can only be of type TIMESTAMP or DATE. The type found is: STRING., invalid
    * google_bigquery_table.telemetry_core_v1: 1 error(s) occurred:

which looks accurate since we're partitioning on `submission_timestamp`.

fbertsch commented:
@whd unfortunate, it doesn't look like we support the `date-time` format in JSON schemas yet in the transpiler. I'll open an issue.

whd commented:
@fbertsch can we also keep all the json schemas in the generated branch i.e. have the allowlist only apply to bq schema generation? This more accurately reflects how the current pipeline is running.

fbertsch commented:
@whd can do, should we keep pioneer schemas?

whd commented:
@fbertsch no need, we can decided later how to handle those.

fbertsch commented:
@whd addressed the feedback. The `submission_timestamp` fields are now `DATETIME`, so we may be able to test deploy.

acmiyaguchi commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,118 @@
+
+find . -type f -name "*.schema.json"|while read fname; do
+    BQ_OUT=${fname/schema.json/bq}
+    jsonschema-transpiler --type bigquery $fname | jq '.fields' > $BQ_OUT
```
This no longer has to be run as of https://github.com/mozilla/jsonschema-transpiler/pull/61, it may actually cause the resulting schemas to be empty.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+
+    rm -rf mozilla-pipeline-schemas
+
+    git clone git@github.com:mozilla-services/mozilla-pipeline-schemas.git
```
Might be good to make the git URL a top-level variable like the rest of the `MPS_` envs.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    # Checkout mozilla-pipeline-schemas and changes directory to prepare for
+    # schema generation.
+
+    rm -rf mozilla-pipeline-schemas
```
Generally I would prefer something like `[[ -d mozilla-pipeline-schemas ]] && rm -rf mozilla-pipeline-schemas` but this is fine.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    rm -rf mozilla-pipeline-schemas
+
+    git clone git@github.com:mozilla-services/mozilla-pipeline-schemas.git
+    cd mozilla-pipeline-schemas/$MPS_SCHEMAS_DIR || exit
```
What is this `|| exit` about?

acmiyaguchi replied:
https://github.com/koalaman/shellcheck/wiki/SC2164

This is to prevent `cd` from the doing the wrong thing during a failure case.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+}
+
+function prepare_metadata() {
+    # Assumes that all schemas under the current directory are valid JSON schema.
```
Might be worth simply adding `-name '*.schema.json`` to the find command(s).

whd commented on bin/allowlist:
```diff
@@ -0,0 +1,5 @@
+telemetry/core
```
This file isn't an executable so in my opinion should not live in `bin`.

acmiyaguchi replied:
I agree, I'll move this to the project root. However, it seems like this should belong in the root of mozilla-pipeline-schemas.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    # Replace backticks with "\|" (can't do that with tr): sed
+    # Find directories that don't match any of the regex expressions: find
+    # Remove them: rm
+    < /app/mozilla-schema-generator/bin/allowlist tr '\n' '`' | \
```
Maybe `$BASE_DIR` again. See previous comment on `allowlist` being a non-binary file. I _think_ this logic can be simplified to:

    find -name '*.bq' | grep -v -f $ALLOWLIST | xargs rm -f

acmiyaguchi replied:
This looks like it does the right thing.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    # Replace newlines with backticks (hard to do with sed): cat | tr
+    # Remove the last backtick; it's the file-ending newline: rev | cut | rev
+    # Replace backticks with "\|" (can't do that with tr): sed
+    # Find directories that don't match any of the regex expressions: find
```
The goal of this step is to remove bq schemas not in `allowlist`, we are keeping the directories and json schemas.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+function main() {
+    cd $BASE_DIR || exit
+
+    # -1. Setup ssh key and git config
```
Why is this -1?

acmiyaguchi replied:
An original comment, I removed the numbering since the functions better define the scope of work.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    prepare_metadata
+
+    # 5. Add transpiled BQ schemas
+    find . -type f -name "*.schema.json"|while read -r fname; do
```
nit: space for pipe

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+}
+
+function main() {
+    cd $BASE_DIR || exit
```
What are these `|| exits` for? There's another one on line 151. We might want to just `set -e` if we're worried about commands failing for various reasons. In fact, we probably do want to `set -e`, but I haven't looked at where that would require changes to the script.

acmiyaguchi replied:
There are a couple of places where the folders and branches are being deleted without checking for existence. I can go back and add `set -eou pipefail` for errors, unset variables, and pipeline errors.

whd commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    # 6. Push to branch of MPS
+    cd ../ || exit
+    commit_schemas
+    git push --force
```
From a technical perspective, force is not required for our use case (as we saw when accidentally committing to dev). I would recommend removing the force. In the rare case where somebody pushes the generated-schemas branch by hand in the middle of the airflow job's execution we would probably want the automated schema push to fail anyway. Maybe we have branch protection or something enabled now that requires this.

acmiyaguchi commented on bin/schema_generator.sh:
```diff
@@ -0,0 +1,156 @@
+    # Installs mozilla-schema-generator in a virtual environment
+
+    virtualenv msg-venv
+    # shellcheck disable=SC1091
```
Shellcheck is a static analysis tool for bash scripts -- this line is to ignore not being able to source the activation script that's being used. 

https://github.com/koalaman/shellcheck/wiki/SC1091

fbertsch commented:
Thanks for picking this up, Anthony! My only input is that the
jsonschema-transpiler dep wont be updated if the install is in the
dockerfile. I would recommend at least checking for an update in the script.

On Wed, May 8, 2019, 2:51 AM Anthony Miyaguchi <notifications@github.com>
wrote:

> *@acmiyaguchi* commented on this pull request.
> ------------------------------
>
> In bin/schema_generator.sh
> <https://github.com/mozilla/mozilla-schema-generator/pull/18#discussion_r281880077>
> :
>
> > +    # Assumes the current directory is the root of the repository
> +
> +    find . -name "*.bq" -type f -exec git add {} +
> +    git checkout ./*.schema.json
> +    git commit -a -m "Interim Commit"
> +
> +    git checkout $MPS_BRANCH_PUBLISH || git checkout -b $MPS_BRANCH_PUBLISH
> +
> +    # Keep only the schemas dir
> +    find .  -mindepth 1 -maxdepth 1 -not -name .git -exec rm -rf {} +
> +    git checkout $MPS_BRANCH_WORKING -- schemas
> +    git commit -a -m "Auto-push from schema generation"
> +}
> +
> +function main() {
> +    cd $BASE_DIR || exit
>
> There are a couple of places where the folders and branches are being
> deleted without checking for existence. I can go back and add set -eou
> pipefail for errors, unset variables, and pipeline errors.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/mozilla-schema-generator/pull/18#discussion_r281880077>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE62YYGQIGHLCQY6NQ774BTPUIPYZANCNFSM4HIQT45A>
> .
>

jklukas commented on bin/allowlist:
```diff
@@ -0,0 +1,5 @@
+org-mozilla-fenix/.*
+org-mozilla-reference-browser/.*
+activity-stream/.*
+eng-workflow/bmobugs
```
```suggestion
eng-workflow/bmobugs
mobile/activation
```

This new ping was just merged in https://github.com/mozilla-services/mozilla-pipeline-schemas/pull/331 and I'd like to start testing it in the gcp pipeline tomorrow.

acmiyaguchi commented on Dockerfile:
```diff
@@ -15,22 +15,37 @@ RUN mkdir ${HOME} && \
+ENV PATH $PATH:$HOME/google-cloud-sdk/bin
+
 # Install Rust and Cargo
 RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain=${RUST_SPEC}
```
Yeah, this is a perfectly legitimate way of installing rust. I've updated it to reflect the default arguments on the website: https://rustup.rs/

I don't think there are going to be any breaking changes to the language in the near future, so using stable should be fine.

acmiyaguchi commented:
> Thanks for picking this up, Anthony! My only input is that the jsonschema-transpiler dep wont be updated if the install is in the dockerfile. I would recommend at least checking for an update in the script.

Unfortunately, there isn't a mechanism for updates. Instead, we would have to just force install over the original. We should just update the tag in the dockerfile when necessary. 

---

@whd r?
